File: sdk/react-native/AuthClient.ts
/**
 * SaaS Auth API - React Native SDK
 * 
 * Comprehensive React Native SDK for authentication and user management.
 * 
 * Features:
 * - User authentication (login, register, logout)
 * - Token management (access, refresh)
 * - OAuth2/OIDC integration
 * - Two-factor authentication
 * - Password reset
 * - Profile management
 * - Real-time notifications via WebSocket
 * - Type-safe with TypeScript
 * - AsyncStorage for token persistence
 * - Secure storage for sensitive data
 */

import AsyncStorage from '@react-native-async-storage/async-storage';
import { Platform } from 'react-native';
import * as SecureStore from 'expo-secure-store';
import { EventRegister } from 'react-native-event-listeners';

// Types
export interface User {
  id: number;
  username: string;
  email: string;
  role: 'user' | 'admin' | 'moderator';
  subscription_plan: 'free' | 'pro' | 'enterprise';
  is_active: boolean;
  email_verified: boolean;
  created_at: string;
  updated_at?: string;
}

export interface AuthTokens {
  access_token: string;
  refresh_token: string;
  token_type: string;
  expires_in: number;
}

export interface LoginCredentials {
  username: string;
  password: string;
}

export interface RegisterData {
  username: string;
  email: string;
  password: string;
}

export interface TwoFactorSetup {
  secret: string;
  qr_code: string;
  backup_codes: string[];
  uri: string;
}

export interface OAuthProvider {
  google: string;
  github: string;
  apple: string;
}

// Configuration
export interface AuthClientConfig {
  baseURL: string;
  timeout?: number;
  enableLogging?: boolean;
  enableWebSocket?: boolean;
  wsURL?: string;
}

// Error types
export class AuthError extends Error {
  constructor(
    message: string,
    public statusCode?: number,
    public code?: string
  ) {
    super(message);
    this.name = 'AuthError';
  }
}

// Response types
export interface ApiResponse<T> {
  data: T;
  message?: string;
  error?: string;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  pages: number;
}

/**
 * Main Auth Client Class
 */
export class AuthClient {
  private config: AuthClientConfig;
  private tokens: AuthTokens | null = null;
  private currentUser: User | null = null;
  private wsConnection: WebSocket | null = null;
  private refreshPromise: Promise<AuthTokens> | null = null;

  // Storage keys
  private static readonly ACCESS_TOKEN_KEY = '@auth_access_token';
  private static readonly REFRESH_TOKEN_KEY = '@auth_refresh_token';
  private static readonly USER_KEY = '@auth_user';

  constructor(config: AuthClientConfig) {
    this.config = {
      timeout: 30000,
      enableLogging: false,
      enableWebSocket: true,
      ...config
    };
  }

  /**
   * Initialize the client and load stored tokens
   */
  async initialize(): Promise<void> {
    try {
      const [accessToken, refreshToken, userStr] = await Promise.all([
        AsyncStorage.getItem(AuthClient.ACCESS_TOKEN_KEY),
        AsyncStorage.getItem(AuthClient.REFRESH_TOKEN_KEY),
        AsyncStorage.getItem(AuthClient.USER_KEY)
      ]);

      if (accessToken && refreshToken) {
        this.tokens = {
          access_token: accessToken,
          refresh_token: refreshToken,
          token_type: 'bearer',
          expires_in: 1800
        };
      }

      if (userStr) {
        this.currentUser = JSON.parse(userStr);
      }

      if (this.config.enableWebSocket && this.tokens) {
        await this.connectWebSocket();
      }
    } catch (error) {
      this.log('Initialization error:', error);
    }
  }

  /**
   * Login with username and password
   */
  async login(credentials: LoginCredentials): Promise<User> {
    const response = await this.request<AuthTokens>('/auth/login', {
      method: 'POST',
      body: JSON.stringify(credentials)
    });

    await this.setTokens(response);
    const user = await this.fetchProfile();
    await this.saveUser(user);
    
    if (this.config.enableWebSocket) {
      await this.connectWebSocket();
    }

    EventRegister.emit('auth:login', user);
    return user;
  }

  /**
   * Register a new user
   */
  async register(data: RegisterData): Promise<User> {
    const response = await this.request<AuthTokens>('/auth/register', {
      method: 'POST',
      body: JSON.stringify(data)
    });

    await this.setTokens(response);
    const user = await this.fetchProfile();
    await this.saveUser(user);
    
    if (this.config.enableWebSocket) {
      await this.connectWebSocket();
    }

    EventRegister.emit('auth:register', user);
    return user;
  }

  /**
   * Logout current user
   */
  async logout(): Promise<void> {
    try {
      await this.request('/auth/logout', { method: 'POST' });
    } catch (error) {
      this.log('Logout request failed:', error);
    }

    await this.clearTokens();
    await this.disconnectWebSocket();
    
    EventRegister.emit('auth:logout');
  }

  /**
   * Refresh access token
   */
  private async refreshTokens(): Promise<AuthTokens> {
    if (this.refreshPromise) {
      return this.refreshPromise;
    }

    this.refreshPromise = (async () => {
      try {
        const response = await this.request<AuthTokens>('/auth/refresh', {
          method: 'POST',
          body: JSON.stringify({
            refresh_token: this.tokens?.refresh_token
          })
        });

        await this.setTokens(response);
        return response;
      } finally {
        this.refreshPromise = null;
      }
    })();

    return this.refreshPromise;
  }

  /**
   * Get current user profile
   */
  async fetchProfile(): Promise<User> {
    return this.request<User>('/users/profile');
  }

  /**
   * Update user profile
   */
  async updateProfile(data: Partial<User>): Promise<User> {
    const user = await this.request<User>('/users/profile', {
      method: 'PUT',
      body: JSON.stringify(data)
    });
    
    await this.saveUser(user);
    return user;
  }

  /**
   * Change password
   */
  async changePassword(oldPassword: string, newPassword: string): Promise<void> {
    await this.request('/auth/change-password', {
      method: 'POST',
      body: JSON.stringify({
        old_password: oldPassword,
        new_password: newPassword
      })
    });
  }

  /**
   * Request password reset
   */
  async requestPasswordReset(email: string): Promise<void> {
    await this.request('/auth/password-reset/request', {
      method: 'POST',
      body: JSON.stringify({ email })
    });
  }

  /**
   * Reset password with token
   */
  async resetPassword(token: string, newPassword: string): Promise<void> {
    await this.request('/auth/password-reset/reset', {
      method: 'POST',
      body: JSON.stringify({
        token,
        new_password: newPassword
      })
    });
  }

  /**
   * Setup two-factor authentication
   */
  async setupTwoFactor(): Promise<TwoFactorSetup> {
    return this.request<TwoFactorSetup>('/auth/2fa/setup', {
      method: 'POST'
    });
  }

  /**
   * Verify two-factor authentication
   */
  async verifyTwoFactor(code: string): Promise<void> {
    await this.request('/auth/2fa/verify', {
      method: 'POST',
      body: JSON.stringify({ code })
    });
  }

  /**
   * Disable two-factor authentication
   */
  async disableTwoFactor(password: string): Promise<void> {
    await this.request('/auth/2fa/disable', {
      method: 'POST',
      body: JSON.stringify({ password })
    });
  }

  /**
   * Get OAuth authorization URL
   */
  async getOAuthUrl(provider: keyof OAuthProvider): Promise<string> {
    const response = await this.request<{ url: string }>(`/auth/oauth/${provider}/url`);
    return response.url;
  }

  /**
   * Handle OAuth callback
   */
  async handleOAuthCallback(provider: keyof OAuthProvider, code: string): Promise<User> {
    const response = await this.request<AuthTokens>(`/auth/oauth/${provider}/callback`, {
      method: 'POST',
      body: JSON.stringify({ code })
    });

    await this.setTokens(response);
    const user = await this.fetchProfile();
    await this.saveUser(user);
    
    if (this.config.enableWebSocket) {
      await this.connectWebSocket();
    }

    return user;
  }

  /**
   * WebSocket connection
   */
  private async connectWebSocket(): Promise<void> {
    if (this.wsConnection) {
      return;
    }

    const wsURL = this.config.wsURL || this.config.baseURL.replace('http', 'ws');
    const token = this.tokens?.access_token;

    if (!token) {
      return;
    }

    this.wsConnection = new WebSocket(`${wsURL}/ws?token=${token}`);

    this.wsConnection.onopen = () => {
      this.log('WebSocket connected');
      EventRegister.emit('ws:connected');
    };

    this.wsConnection.onmessage = (event) => {
      try {
        const message = JSON.parse(event.data);
        EventRegister.emit('ws:message', message);
      } catch (error) {
        this.log('WebSocket message parse error:', error);
      }
    };

    this.wsConnection.onerror = (error) => {
      this.log('WebSocket error:', error);
      EventRegister.emit('ws:error', error);
    };

    this.wsConnection.onclose = () => {
      this.log('WebSocket closed');
      this.wsConnection = null;
      EventRegister.emit('ws:disconnected');
      
      // Attempt reconnection after delay
      setTimeout(() => {
        if (this.tokens) {
          this.connectWebSocket();
        }
      }, 5000);
    };
  }

  /**
   * Disconnect WebSocket
   */
  private async disconnectWebSocket(): Promise<void> {
    if (this.wsConnection) {
      this.wsConnection.close();
      this.wsConnection = null;
    }
  }

  /**
   * Send WebSocket message
   */
  sendWebSocketMessage(type: string, data: any): void {
    if (this.wsConnection && this.wsConnection.readyState === WebSocket.OPEN) {
      this.wsConnection.send(JSON.stringify({ type, data }));
    }
  }

  /**
   * Make authenticated API request
   */
  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const url = `${this.config.baseURL}${endpoint}`;
    
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      ...options.headers
    };

    if (this.tokens?.access_token) {
      headers['Authorization'] = `Bearer ${this.tokens.access_token}`;
    }

    try {
      let response = await fetch(url, {
        ...options,
        headers,
        signal: AbortSignal.timeout(this.config.timeout!)
      });

      // Handle 401 Unauthorized - try refresh
      if (response.status === 401 && this.tokens?.refresh_token) {
        await this.refreshTokens();
        headers['Authorization'] = `Bearer ${this.tokens.access_token}`;
        response = await fetch(url, {
          ...options,
          headers,
          signal: AbortSignal.timeout(this.config.timeout!)
        });
      }

      if (!response.ok) {
        const error = await response.json().catch(() => ({ detail: 'Request failed' }));
        throw new AuthError(
          error.detail || 'Request failed',
          response.status,
          error.code
        );
      }

      return response.json();
    } catch (error) {
      if (error instanceof AuthError) {
        throw error;
      }
      throw new AuthError(
        error instanceof Error ? error.message : 'Network error',
        undefined,
        'NETWORK_ERROR'
      );
    }
  }

  /**
   * Set and store tokens
   */
  private async setTokens(tokens: AuthTokens): Promise<void> {
    this.tokens = tokens;
    
    await Promise.all([
      AsyncStorage.setItem(AuthClient.ACCESS_TOKEN_KEY, tokens.access_token),
      AsyncStorage.setItem(AuthClient.REFRESH_TOKEN_KEY, tokens.refresh_token)
    ]);
  }

  /**
   * Clear stored tokens
   */
  private async clearTokens(): Promise<void> {
    this.tokens = null;
    this.currentUser = null;
    
    await Promise.all([
      AsyncStorage.removeItem(AuthClient.ACCESS_TOKEN_KEY),
      AsyncStorage.removeItem(AuthClient.REFRESH_TOKEN_KEY),
      AsyncStorage.removeItem(AuthClient.USER_KEY)
    ]);
  }

  /**
   * Save user data
   */
  private async saveUser(user: User): Promise<void> {
    this.currentUser = user;
    await AsyncStorage.setItem(AuthClient.USER_KEY, JSON.stringify(user));
  }

  /**
   * Get current user
   */
  getCurrentUser(): User | null {
    return this.currentUser;
  }

  /**
   * Check if user is authenticated
   */
  isAuthenticated(): boolean {
    return !!this.tokens?.access_token;
  }

  /**
   * Get access token
   */
  getAccessToken(): string | null {
    return this.tokens?.access_token || null;
  }

  /**
   * Logging helper
   */
  private log(...args: any[]): void {
    if (this.config.enableLogging) {
      console.log('[AuthClient]', ...args);
    }
  }
}

/**
 * React Hook for authentication
 */
import { useState, useEffect, useCallback } from 'react';

export function useAuth(client: AuthClient) {
  const [user, setUser] = useState<User | null>(client.getCurrentUser());
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<AuthError | null>(null);

  useEffect(() => {
    client.initialize().then(() => {
      setUser(client.getCurrentUser());
      setLoading(false);
    });

    const loginListener = EventRegister.addEventListener('auth:login', (loggedInUser) => {
      setUser(loggedInUser);
    });

    const logoutListener = EventRegister.addEventListener('auth:logout', () => {
      setUser(null);
    });

    return () => {
      EventRegister.removeEventListener(loginListener);
      EventRegister.removeEventListener(logoutListener);
    };
  }, [client]);

  const login = useCallback(async (credentials: LoginCredentials) => {
    setLoading(true);
    setError(null);
    try {
      const loggedInUser = await client.login(credentials);
      setUser(loggedInUser);
      return loggedInUser;
    } catch (e) {
      setError(e as AuthError);
      throw e;
    } finally {
      setLoading(false);
    }
  }, [client]);

  const register = useCallback(async (data: RegisterData) => {
    setLoading(true);
    setError(null);
    try {
      const registeredUser = await client.register(data);
      setUser(registeredUser);
      return registeredUser;
    } catch (e) {
      setError(e as AuthError);
      throw e;
    } finally {
      setLoading(false);
    }
  }, [client]);

  const logout = useCallback(async () => {
    setLoading(true);
    try {
      await client.logout();
      setUser(null);
    } catch (e) {
      setError(e as AuthError);
    } finally {
      setLoading(false);
    }
  }, [client]);

  return {
    user,
    loading,
    error,
    isAuthenticated: client.isAuthenticated(),
    login,
    register,
    logout,
    client
  };
}

/**
 * React Hook for WebSocket
 */
export function useWebSocket(client: AuthClient) {
  const [connected, setConnected] = useState(false);
  const [messages, setMessages] = useState<any[]>([]);

  useEffect(() => {
    const connectedListener = EventRegister.addEventListener('ws:connected', () => {
      setConnected(true);
    });

    const disconnectedListener = EventRegister.addEventListener('ws:disconnected', () => {
      setConnected(false);
    });

    const messageListener = EventRegister.addEventListener('ws:message', (message) => {
      setMessages(prev => [...prev, message]);
    });

    return () => {
      EventRegister.removeEventListener(connectedListener);
      EventRegister.removeEventListener(disconnectedListener);
      EventRegister.removeEventListener(messageListener);
    };
  }, []);

  const sendMessage = useCallback((type: string, data: any) => {
    client.sendWebSocketMessage(type, data);
  }, [client]);

  return {
    connected,
    messages,
    sendMessage
  };
}

// Export default
export default AuthClient;

File: sdk/flutter/auth_client.dart
/// SaaS Auth API - Flutter SDK
///
/// Comprehensive Flutter SDK for authentication and user management.
///
/// Features:
/// - User authentication (login, register, logout)
/// - Token management (access, refresh)
/// - OAuth2/OIDC integration
/// - Two-factor authentication
/// - Password reset
/// - Profile management
/// - Real-time notifications via WebSocket
/// - Type-safe with Dart
/// - Secure storage for tokens
/// - State management integration

library auth_client;

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:connectivity_plus/connectivity_plus.dart';

// ==================== Models ====================

/// User model
class User {
  final int id;
  final String username;
  final String email;
  final String role;
  final String subscriptionPlan;
  final bool isActive;
  final bool emailVerified;
  final DateTime createdAt;
  final DateTime? updatedAt;

  User({
    required this.id,
    required this.username,
    required this.email,
    required this.role,
    required this.subscriptionPlan,
    required this.isActive,
    required this.emailVerified,
    required this.createdAt,
    this.updatedAt,
  });

  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      id: json['id'],
      username: json['username'],
      email: json['email'],
      role: json['role'],
      subscriptionPlan: json['subscription_plan'] ?? 'free',
      isActive: json['is_active'] ?? false,
      emailVerified: json['email_verified'] ?? false,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );

    Map<String, dynamic> toJson() {
      return {
        'id': id,
        'username': username,
        'email': email,
        'role': role,
        'subscription_plan': subscriptionPlan,
        'is_active': isActive,
        'email_verified': emailVerified,
        'created_at': createdAt.toIso8601String(),
        'updated_at': updatedAt?.toIso8601String(),
      };
    }
  }

  /// Auth tokens model
  class AuthTokens {
    final String accessToken;
    final String refreshToken;
    final String tokenType;
    final int expiresIn;

    AuthTokens({
      required this.accessToken,
      required this.refreshToken,
      required this.tokenType,
      required this.expiresIn,
    });

    factory AuthTokens.fromJson(Map<String, dynamic> json) {
      return AuthTokens(
        accessToken: json['access_token'],
        refreshToken: json['refresh_token'],
        tokenType: json['token_type'] ?? 'bearer',
        expiresIn: json['expires_in'] ?? 1800,
      );
    }

    Map<String, dynamic> toJson() {
      return {
        'access_token': accessToken,
        'refresh_token': refreshToken,
        'token_type': tokenType,
        'expires_in': expiresIn,
      };
    }
  }

  /// Login credentials
  class LoginCredentials {
    final String username;
    final String password;

    LoginCredentials({
      required this.username,
      required this.password,
    });

    Map<String, dynamic> toJson() {
      return {
        'username': username,
        'password': password,
      };
    }
  }

  /// Register data
  class RegisterData {
    final String username;
    final String email;
    final String password;

    RegisterData({
      required this.username,
      required this.email,
      required this.password,
    });

    Map<String, dynamic> toJson() {
      return {
        'username': username,
        'email': email,
        'password': password,
      };
    }
  }

  /// Two-factor setup data
  class TwoFactorSetup {
    final String secret;
    final String qrCode;
    final List<String> backupCodes;
    final String uri;

    TwoFactorSetup({
      required this.secret,
      required this.qrCode,
      required this.backupCodes,
      required this.uri,
    });

    factory TwoFactorSetup.fromJson(Map<String, dynamic> json) {
      return TwoFactorSetup(
        secret: json['secret'],
        qrCode: json['qr_code'],
        backupCodes: List<String>.from(json['backup_codes']),
        uri: json['uri'],
      );
    }
  }

  /// API response wrapper
  class ApiResponse<T> {
    final T? data;
    final String? message;
    final String? error;

    ApiResponse({
      this.data,
    this.message,
    this.error,
  });

    factory ApiResponse.fromJson(
      Map<String, dynamic> json,
      T Function(Map<String, dynamic>) fromJsonT,
      ) {
      return ApiResponse<T>(
        data: json['data'] != null ? fromJsonT(json['data']) : null,
        message: json['message'],
        error: json['error'],
      );
    }
  }

  /// Paginated response
  class PaginatedResponse<T> {
    final List<T> items;
    final int total;
    final int page;
    final int pages;

    PaginatedResponse({
      required this.items,
      required this.total,
    required this.page,
    required this.pages,
  });

    factory PaginatedResponse.fromJson(
      Map<String, dynamic> json,
      T Function(Map<String, dynamic>) fromJsonT,
      ) {
    return PaginatedResponse<T>(
      items: (json['items'] as List)
          .map((item) => fromJsonT(item))
          .toList(),
      total: json['total'],
      page: json['page'],
      pages: json['pages'],
    );
  }
}

// ==================== Configuration ====================

/// Auth client configuration
class AuthClientConfig {
  final String baseURL;
  final Duration timeout;
  final bool enableLogging;
  final bool enableWebSocket;
  final String? wsURL;

  AuthClientConfig({
    required this.baseURL,
    this.timeout = const Duration(seconds: 30),
    this.enableLogging = false,
    this.enableWebSocket = true,
    this.wsURL,
  });
}

// ==================== Exceptions ====================

/// Auth error
class AuthError implements Exception {
  final String message;
  final int? statusCode;
  final String? code;

  AuthError({
    required this.message,
    this.statusCode,
    this.code,
  });

  @override
  String toString() => 'AuthError: $message (code: $code, status: $statusCode)';
}

// ==================== Main Client ====================

/// Main Auth Client
class AuthClient {
  final AuthClientConfig config;
  final http.Client _httpClient;
  AuthTokens? _tokens;
  User? _currentUser;
  WebSocketChannel? _wsChannel;
  StreamController<dynamic>? _messageController;
  Future<AuthTokens>? _refreshPromise;

  // Storage keys
  static const String _accessTokenKey = 'auth_access_token';
  static const String _refreshTokenKey = 'auth_refresh_token';
  static const String _userKey = 'auth_user';

  // Event streams
  final StreamController<User> _authEventController = StreamController<User>.broadcast();
  final StreamController<String> _wsEventController = StreamController<String>.broadcast();

  AuthClient({
    required this.config,
    http.Client? httpClient,
  }) : _httpClient = httpClient ?? http.Client();

  /// Get auth event stream
  Stream<User> get authEvents => _authEventController.stream;

  /// Get WebSocket event stream
  Stream<String> get wsEvents => _wsEventController.stream;

  /// Get current user
  User? get currentUser => _currentUser;

  /// Check if authenticated
  bool get isAuthenticated => _tokens?.accessToken != null;

  /// Get access token
  String? get accessToken => _tokens?.accessToken;

  /// Initialize the client
  Future<void> initialize() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final accessToken = prefs.getString(_accessTokenKey);
      final refreshToken = prefs.getString(_refreshTokenKey);
      final userStr = prefs.getString(_userKey);

      if (accessToken != null && refreshToken != null) {
        _tokens = AuthTokens(
          accessToken: accessToken,
          refreshToken: refreshToken,
          tokenType: 'bearer',
          expiresIn: 1800,
        );
      }

      if (userStr != null) {
        _currentUser = User.fromJson(jsonDecode(userStr));
      }

      if (config.enableWebSocket && _tokens != null) {
        await connectWebSocket();
      }
    } catch (e) {
      _log('Initialization error: $e');
    }
  }

  /// Login with credentials
  Future<User> login(LoginCredentials credentials) async {
    final response = await _request<AuthTokens>(
      '/auth/login',
      method: 'POST',
      body: jsonEncode(credentials.toJson()),
    );

    await _setTokens(response);
    final user = await fetchProfile();
    await _saveUser(user);

    if (config.enableWebSocket) {
      await connectWebSocket();
    }

    _authEventController.add(user);
    return user;
  }

  /// Register new user
  Future<User> register(RegisterData data) async {
    final response = await _request<AuthTokens>(
      '/auth/register',
      method: 'POST',
      body: jsonEncode(data.toJson()),
    );

    await _setTokens(response);
    final user = await fetchProfile();
    await _saveUser(user);

    if (config.enableWebSocket) {
      await connectWebSocket();
    }

    _authEventController.add(user);
    return user;
  }

  /// Logout
  Future<void> logout() async {
    try {
      await _request('/auth/logout', method: 'POST');
    } catch (e) {
      _log('Logout request failed: $e');
    }

    await _clearTokens();
    await disconnectWebSocket();
    _authEventController.add(User(
      id: 0,
      username: '',
      email: '',
      role: '',
      subscriptionPlan: 'free',
      isActive: false,
      emailVerified: false,
      createdAt: DateTime.now(),
    ));
  }

  /// Refresh tokens
  Future<AuthTokens> _refreshTokens() async {
    if (_refreshPromise != null) {
      return _refreshPromise!;
    }

    _refreshPromise = (() async {
      try {
        final response = await _request<AuthTokens>(
          '/auth/refresh',
          method: 'POST',
          body: jsonEncode({
            'refresh_token': _tokens?.refreshToken,
          }),
        );

        await _setTokens(response);
        return response;
      } finally {
        _refreshPromise = null;
      }
    })();

    return _refreshPromise!;
  }

  /// Fetch user profile
  Future<User> fetchProfile() async {
    return _request<User>('/users/profile');
  }

  /// Update user profile
  Future<User> updateProfile(Map<String, dynamic> data) async {
    final user = await _request<User>(
      '/users/profile',
      method: 'PUT',
      body: jsonEncode(data),
    );

    await _saveUser(user);
    return user;
  }

  /// Change password
  Future<void> changePassword(String oldPassword, String newPassword) async {
    await _request(
      '/auth/change-password',
      method: 'POST',
      body: jsonEncode({
        'old_password': oldPassword,
        'new_password': newPassword,
      }),
    );
  }

  /// Request password reset
  Future<void> requestPasswordReset(String email) async {
    await _request(
      '/auth/password-reset/request',
      method: 'POST',
      body: jsonEncode({'email': email}),
    );
  }

  /// Reset password with token
  Future<void> resetPassword(String token, String newPassword) async {
    await _request(
      '/auth/password-reset/reset',
      method: 'POST',
      body: jsonEncode({
        'token': token,
        'new_password': newPassword,
      }),
    );
  }

  /// Setup two-factor authentication
  Future<TwoFactorSetup> setupTwoFactor() async {
    return _request<TwoFactorSetup>(
      '/auth/2fa/setup',
      method: 'POST',
    );
  }

  /// Verify two-factor authentication
  Future<void> verifyTwoFactor(String code) async {
    await _request(
      '/auth/2fa/verify',
      method: 'POST',
      body: jsonEncode({'code': code}),
    );
  }

  /// Disable two-factor authentication
  Future<void> disableTwoFactor(String password) async {
    await _request(
      '/auth/2fa/disable',
      method: 'POST',
      body: jsonEncode({'password': password}),
    );
  }

  /// Get OAuth URL
  Future<String> getOAuthUrl(String provider) async {
    final response = await _request<Map<String, dynamic>>(
      '/auth/oauth/$provider/url',
    );
    return response['url'];
  }

  /// Handle OAuth callback
  Future<User> handleOAuthCallback(String provider, String code) async {
    final response = await _request<AuthTokens>(
      '/auth/oauth/$provider/callback',
      method: 'POST',
      body: jsonEncode({'code': code}),
    );

    await _setTokens(response);
    final user = await fetchProfile();
    await _saveUser(user);

    if (config.enableWebSocket) {
      await connectWebSocket();
    }

    return user;
  }

  /// Connect WebSocket
  Future<void> connectWebSocket() async {
    if (_wsChannel != null) {
      return;
    }

    final wsURL = config.wsURL ?? config.baseURL.replace('http', 'ws');
    final token = _tokens?.accessToken;

    if (token == null) {
      return;
    }

    try {
      _wsChannel = WebSocketChannel.connect(
        Uri.parse('$wsURL/ws?token=$token'),
      );

      _messageController = StreamController<dynamic>.broadcast();

      _wsChannel!.stream.listen(
        (message) {
          _log('WebSocket message: $message');
          _wsEventController.add(message.toString());
          _messageController?.add(message);
        },
        onError: (error) {
          _log('WebSocket error: $error');
          _wsEventController.add('error:$error');
        },
        onDone: () {
          _log('WebSocket closed');
          _wsChannel = null;
          _wsEventController.add('disconnected');
          
          // Attempt reconnection
          Future.delayed(const Duration(seconds: 5), () {
            if (_tokens != null) {
              connectWebSocket();
            }
          });
        },
      );

      _log('WebSocket connected');
      _wsEventController.add('connected');
    } catch (e) {
      _log('WebSocket connection error: $e');
    }
  }

  /// Disconnect WebSocket
  Future<void> disconnectWebSocket() async {
    await _wsChannel?.sink.close();
    _wsChannel = null;
    await _messageController?.close();
    _messageController = null;
  }

  /// Send WebSocket message
  void sendWebSocketMessage(String type, dynamic data) {
    if (_wsChannel != null) {
      _wsChannel!.sink.add(jsonEncode({'type': type, 'data': data}));
    }
  }

  /// Get WebSocket message stream
  Stream<dynamic>? get wsMessages => _messageController?.stream;

  /// Make authenticated API request
  Future<T> _request<T>(
    String endpoint, {
    String method = 'GET',
    String? body,
    Map<String, String>? headers,
  }) async {
    final url = Uri.parse('${config.baseURL}$endpoint');
    final requestHeaders = <String, String>{
      'Content-Type': 'application/json',
      ...?headers,
    };

    if (_tokens?.accessToken != null) {
      requestHeaders['Authorization'] = 'Bearer ${_tokens!.accessToken}';
    }

    http.Response response;

    try {
      // Check connectivity
      final connectivityResult = await Connectivity().checkConnectivity();
      if (connectivityResult == ConnectivityResult.none) {
        throw AuthError(
          message: 'No internet connection',
          code: 'NO_CONNECTION',
        );
      }

      response = await _httpClient
          .send(http.Request(method, url))
          .timeout(config.timeout)
          .then((req) async {
        if (body != null) {
          req.body = body;
        }
        req.headers.addAll(requestHeaders);
        return await _httpClient.send(req);
      });

      // Handle 401 - try refresh
      if (response.statusCode == 401 && _tokens?.refreshToken != null) {
        await _refreshTokens();
        requestHeaders['Authorization'] = 'Bearer ${_tokens!.accessToken}';
        response = await _httpClient
            .send(http.Request(method, url))
            .timeout(config.timeout)
            .then((req) async {
          if (body != null) {
            req.body = body;
          }
          req.headers.addAll(requestHeaders);
          return await _httpClient.send(req);
        });
      }

      if (response.statusCode >= 400) {
        final errorData = jsonDecode(response.body);
        throw AuthError(
          message: errorData['detail'] ?? 'Request failed',
          statusCode: response.statusCode,
          code: errorData['code'],
        );
      }

      final responseData = jsonDecode(response.body);

      if (T == User) {
        return User.fromJson(responseData) as T;
      } else if (T == AuthTokens) {
        return AuthTokens.fromJson(responseData) as T;
      } else if (T == TwoFactorSetup) {
        return TwoFactorSetup.fromJson(responseData) as T;
      } else {
        return responseData as T;
      }
    } on SocketException catch (e) {
      throw AuthError(
        message: 'Network error: ${e.message}',
        code: 'NETWORK_ERROR',
      );
    } on TimeoutException catch (e) {
      throw AuthError(
        message: 'Request timeout',
        code: 'TIMEOUT',
      );
    } on AuthError {
      rethrow;
    } catch (e) {
      throw AuthError(
        message: e.toString(),
        code: 'UNKNOWN_ERROR',
      );
    }
  }

  /// Set and store tokens
  Future<void> _setTokens(AuthTokens tokens) async {
    _tokens = tokens;
    final prefs = await SharedPreferences.getInstance();
    await Future.wait([
      prefs.setString(_accessTokenKey, tokens.accessToken),
      prefs.setString(_refreshTokenKey, tokens.refreshToken),
    ]);
  }

  /// Clear stored tokens
  Future<void> _clearTokens() async {
    _tokens = null;
    _currentUser = null;
    final prefs = await SharedPreferences.getInstance();
    await Future.wait([
      prefs.remove(_accessTokenKey),
      prefs.remove(_refreshTokenKey),
      prefs.remove(_userKey),
    ]);
  }

  /// Save user data
  Future<void> _saveUser(User user) async {
    _currentUser = user;
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_userKey, jsonEncode(user.toJson()));
  }

  /// Logging helper
  void _log(String message) {
    if (config.enableLogging) {
      debugPrint('[AuthClient] $message');
    }
  }

  /// Dispose resources
  void dispose() {
    disconnectWebSocket();
    _authEventController.close();
    _wsEventController.close();
    _httpClient.close();
  }
}

// ==================== Provider ====================

/// Auth state provider for Flutter
import 'package:flutter/material.dart';

class AuthProvider extends ChangeNotifier {
  final AuthClient client;
  User? _user;
  bool _loading = true;
  AuthError? _error;

  AuthProvider({required this.client}) {
    _initialize();
  }

  User? get user => _user;
  bool get loading => _loading;
  AuthError? get error => _error;
  bool get isAuthenticated => client.isAuthenticated;

  Future<void> _initialize() async {
    await client.initialize();
    _user = client.currentUser;
    _loading = false;
    notifyListeners();

    client.authEvents.listen((user) {
      _user = user;
      notifyListeners();
    });
  }

  Future<User> login(LoginCredentials credentials) async {
    _loading = true;
    _error = null;
    notifyListeners();

    try {
      final user = await client.login(credentials);
      _user = user;
      _loading = false;
      notifyListeners();
      return user;
    } on AuthError catch (e) {
      _error = e;
      _loading = false;
      notifyListeners();
      rethrow;
    }
  }

  Future<User> register(RegisterData data) async {
    _loading = true;
    _error = null;
    notifyListeners();

    try {
      final user = await client.register(data);
      _user = user;
      _loading = false;
      notifyListeners();
      return user;
    } on AuthError catch (e) {
      _error = e;
      _loading = false;
      notifyListeners();
      rethrow;
    }
  }

  Future<void> logout() async {
    _loading = true;
    notifyListeners();

    try {
      await client.logout();
      _user = null;
      _loading = false;
      notifyListeners();
    } on AuthError catch (e) {
      _error = e;
      _loading = false;
      notifyListeners();
    }
  }

  @override
  void dispose() {
    client.dispose();
    super.dispose();
  }
}
